### PR TITLE
Sqs processing

### PR DESCRIPTION
**Pull request by maxinelasp**
# Change Summary
Added a queue in front of batch_processing per our discussion about eliminating database race conditions. This did involve an update to batch_starter to change the events getting passed in, but beyond that there are minimal impacts.

## Overview

Rather than eventbridge events directly triggering batch_starter, the events instead go into a queue. 

This is a FIFO queue with multiple message groups. Right now, the message groups are divided by instrument, but they could be further divided. Within the group, they are handled in a FIFO manner, meaning they will be processed in order and processing won't proceed until each event is complete. However, multiple groups are processed in parallel. 

I created an eventbridge rule for each instrument that sorts the events into the queue message groups. This required adding an additional piece of information to each event - instrument name. 

Then, the target for the batch_starter is switched from eventbridge to lambda and everything else is handled.

Notes:
- SQS by default has some batching properties, so I updated batch_starter to loop through each event in the batch
- I looked into having one eventbridge rule for the queue, but I couldn't figure out how to extract the message detail of the event
- Adding additional message groups will create increased parallelism and therefore increase processing speed. Due to the batching, I don't think this will be necessary. 
- We could turn off batching but I like it in this case, it keeps the lambdas warm

## New Files

- sqs_stack.py
   - Creates SQS queue and eventbridge rules

## Updated Files

- BatchStarterStack
   - changed source target from eventbridge to SQS
- batch_starter.py
   - Changed input to a list of SQS events

### CODE REVIEW

File: sds_data_manager/stacks/sqs_stack.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+"""Stack for creating instrument SQS queues and attaching batch_starter as a target."""
+
+from aws_cdk import Duration, Stack, aws_sqs
+from aws_cdk import aws_events as events
+from aws_cdk import aws_events_targets as targets
+from constructs import Construct
+
+
+class SqsStack(Stack):
```
Comment (laspsandoval) [inline sds_data_manager/stacks/sqs_stack.py:9]: I vaguely remember that it is best practice to use a dead letter queue for SQS. Are you planning on adding that in?

File: sds_data_manager/stacks/sqs_stack.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        # This needs to be a FIFO queue to enforce ordering
+        self.instrument_queue = aws_sqs.Queue(
+            self,
+            "FileArrivalQueue",
+            queue_name="file_arrival_queue.fifo",
+            fifo=True,
+            encryption=aws_sqs.QueueEncryption.UNENCRYPTED,
+            # This timeout determines how long the queue waits for processing. It must
+            # be longer than the timeout of the lambda.
+            visibility_timeout=Duration.seconds(300),
```
Comment (laspsandoval) [inline sds_data_manager/stacks/sqs_stack.py:53]: Add similar comment in the lambda. What would happen if someone accidentally set the lambda to be greater than 300s? Would the stack fail to deploy or go unnoticed?

File: sds_data_manager/stacks/sqs_stack.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            # This is required. It removes messages with identical content. Since
+            # the event includes a filename each event should be totally unique.
+            content_based_deduplication=True,
+        )
+
+        for instrument_name in instrument_names:
```
Comment (laspsandoval) [inline sds_data_manager/stacks/sqs_stack.py:63]: How will this work for SPICE? Maybe we aren't there yet, but something to think about for next PR.

Comment (laspsandoval) [reply sds_data_manager/stacks/sqs_stack.py:63]: When we are reprocessing will SQS FIFO slow us way down if we don't have the date as part of the event Rule?

File: sds_data_manager/lambda_code/SDSCode/batch_starter.py [hunk 4]
```diff
@@ -360,20 +361,27 @@ def try_to_submit_job(session, job_info):
...
-        potential_jobs = get_downstream_dependencies(session, components)
-        logger.info(f"Potential jobs found [{len(potential_jobs)}]: {potential_jobs}")
+        filename = body["detail"]["object"]["key"]
+        logger.info(f"Retrieved filename: {filename}")
+        components = ScienceFilePath.extract_filename_components(filename)
+        logger.info(f"Initial event parsed filename: {components}")
 
-        for job in potential_jobs:
-            try_to_submit_job(session, job)
+        with db.Session() as session:
```
Comment (greglucas) [inline sds_data_manager/lambda_code/SDSCode/batch_starter.py]: Maybe bring this outside of the event for-loop too so you're only grabbing one session for each Lambda invocation?

File: sds_data_manager/stacks/indexer_lambda_stack.py [hunk 5]
```diff
@@ -122,6 +122,7 @@ def __init__(
             ),
         )
 
+        # ADD SQS HERE
```
Comment (greglucas) [inline sds_data_manager/stacks/indexer_lambda_stack.py]: Do you still intend to add something here, or unneeded now?

**Comment by laspsandoval** [comment]
Great start. I would like to discuss how it works only using instrument name and reprocessing.

File: sds_data_manager/stacks/sqs_stack.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            Parent construct.
+        construct_id : str
+            A unique string identifier for this construct.
+        instrument_names : list[str]
+            A list of all instrument names
+        kwargs : dict
+            Keyword arguments
+        """
+        super().__init__(scope, construct_id, **kwargs)
+        # This needs to be a FIFO queue to enforce ordering
```
Comment (greglucas) [inline sds_data_manager/stacks/sqs_stack.py:44]: Out of curiosity, do we need to enforce ordering? I think our checks say "if all dependencies are met continue on", which is independent on which incoming file triggered the job.

File: sds_data_manager/stacks/sqs_stack.py (hunk 2 above)
Comment (greglucas) [inline sds_data_manager/stacks/sqs_stack.py:53]: I think this is a re-do timer, not the initial timer? i.e. we aren't waiting 5 minutes here after a file arrives are we? Suggest rewording comment. Also the 30 second default seems like it should be reasonable as well and not have to go down to 5 minutes here or is our database kickoff lambda slow? (I figured it was a few seconds)

**Review by laspsandoval — COMMENTED** [review]
Some comments mostly for me to talk it out with you.

**Comment by greglucas** [comment]
I'm still not convinced that this is absolutely necessary and to me it still adds complexity to the system with another resource to keep track of. But, it does look fine, and no major complaints from me.

File: sds_data_manager/stacks/sqs_stack.py (hunk 1 above)
Comment (maxinelasp) [reply sds_data_manager/stacks/sqs_stack.py:9]: I wasn't planning on it, but I am happy to if we would like that extra bit of resiliency/monitoring.

File: sds_data_manager/stacks/sqs_stack.py (hunk 2 above)
Comment (maxinelasp) [reply sds_data_manager/stacks/sqs_stack.py:53]: The stack does fail to deploy

File: sds_data_manager/stacks/sqs_stack.py (hunk 3 above)
Comment (maxinelasp) [reply sds_data_manager/stacks/sqs_stack.py:63]: It will slow processing down. I would be surprised if it ended up causing issues, but we can also change settings to reduce that issue (for example by batching processing together to eliminate lambda startup times) or add additional queues

File: sds_data_manager/stacks/indexer_lambda_stack.py (hunk 5 above)
Comment (maxinelasp) [reply sds_data_manager/stacks/indexer_lambda_stack.py]: Oops, old comment, I will remove

File: sds_data_manager/stacks/sqs_stack.py (hunk 6 above)
Comment (maxinelasp) [reply sds_data_manager/stacks/sqs_stack.py:44]: We don't need to enforce ordering from an abstract perspective. However, in order to enforce concurrency requirements, AWS requires a FIFO queue. So, in order to enforce the one queue per processing step rule it does need to be FIFO.

File: sds_data_manager/stacks/sqs_stack.py (hunk 2 above)
Comment (maxinelasp) [reply sds_data_manager/stacks/sqs_stack.py:53]: No, we aren't waiting 5 minutes for a file arrives. It's a timeout for how long the queue will wait before moving the event to the dead letter queue, I believe. I made it long just to accommodate any changes in the future to that lambda - it only comes into play if processing is stalled for some reason, so I made it longer just in case. In most cases, processing takes a second or two and the events are removed from the queue that quickly.

File: sds_data_manager/stacks/sqs_stack.py (hunk 3 above)
Comment (maxinelasp) [reply sds_data_manager/stacks/sqs_stack.py:63]: Ultimately this is not going to be a bottleneck for processing because the actual processing steps still run in parallel (with a few seconds delayed starting) and they will take much longer than triggering jobs will.

File: sds_data_manager/stacks/instrument_lambdas.py [hunk 7]
```diff
@@ -102,20 +105,8 @@ def __init__(
...
-                detail={
-                    "object": {"key": [{"exists": True}]},
-                },
-            ),
-        )
-
-        event_from_indexer_lambda.add_target(
-            targets.LambdaFunction(self.instrument_lambda)
-        )
+        # Point the provided instrument queue to trigger the lambda. This will
```
Comment (tech3371) [inline sds_data_manager/stacks/instrument_lambdas.py]: I think I am little confused with this comment. Can you clarify it? Thank you!

File: sds_data_manager/utils/stackbuilder.py [hunk 8]
```diff
@@ -148,6 +149,14 @@ def build_sds(
             env=env,
         )
 
+    # Create SQS pipeline for each instrument and add it to instrument_sqs
```
Comment (tech3371) [inline sds_data_manager/utils/stackbuilder.py:152]: Maxine, can you check if I added architecture changes from this PR in our architecture diagram correctly? Appreciate it!
https://lasp.colorado.edu/galaxy/display/IMAP/IMAP+EventBridge+Rules

File: sds_data_manager/lambda_code/SDSCode/batch_starter.py [hunk 9]
```diff
@@ -360,20 +361,27 @@ def try_to_submit_job(session, job_info):
     logger.info(f"Submitted job {job_name} with this command: {batch_command}")
 
 
-def lambda_handler(event: dict, context):
+def lambda_handler(events: dict, context):
     """Lambda handler."""
-    logger.info(f"Event: {event}")
+    logger.info(f"Event: {events}")
```
Comment (tech3371) [inline sds_data_manager/lambda_code/SDSCode/batch_starter.py]:
```suggestion
    logger.info(f"Events: {events}")
```

File: sds_data_manager/lambda_code/SDSCode/batch_starter.py [hunk 10]
```diff
@@ -360,20 +361,27 @@ def try_to_submit_job(session, job_info):
...
     logger.info(f"Context: {context}")
 
-    # Event details:
-    filename = event["detail"]["object"]["key"]
-    components = ScienceFilePath.extract_filename_components(filename)
-    logger.info(f"Initial event parsed filename: {components}")
+    for event in events["Records"]:
```
Comment (tech3371) [inline sds_data_manager/lambda_code/SDSCode/batch_starter.py]: can you add this comment from your PR description here? I think it will be helpful later.
```
SQS by default has some batching properties, so I updated batch_starter to loop through each event in the batch
```

**Review by tech3371 — COMMENTED** [review]
couple minor questions. Otherwise, it looks good to me.

File: sds_data_manager/stacks/instrument_lambdas.py (hunk 7 above)
Comment (maxinelasp) [reply sds_data_manager/stacks/instrument_lambdas.py]: Ah yes, this is slightly outdated. I will fix it, thanks for pointing it out!

File: sds_data_manager/utils/stackbuilder.py (hunk 8 above)
Comment (maxinelasp) [reply sds_data_manager/utils/stackbuilder.py:152]: Currently, the lambda putevent used by the indexer lambda template only has filename and instrument, not data_level or data_product. I'm not sure if that's the same product as the one in your document. The diagram looks good to me!

File: sds_data_manager/lambda_code/SDSCode/batch_starter.py (hunk 10 above)
Comment (maxinelasp) [reply sds_data_manager/lambda_code/SDSCode/batch_starter.py]: Sure, will do

File: sds_data_manager/utils/stackbuilder.py (hunk 8 above)
Comment (tech3371) [reply sds_data_manager/utils/stackbuilder.py:152]: Yeah, that put event was triggering batch starter lambda before and only passed filename. Now with your changes, it has instrument name as well and also SQS between that putevent and batch starter lambda. That's what I understood from this PR.

Cool, thank you for confirming!

**Comment by maxinelasp** [comment]
@tech3371 @laspsandoval I addressed your comments and added a DLQ, could I get another review please? 😁

File: sds_data_manager/lambda_code/SDSCode/batch_starter.py [hunk 11]
```diff
@@ -360,20 +361,28 @@ def try_to_submit_job(session, job_info):
...
-    # Event details:
-    filename = event["detail"]["object"]["key"]
-    components = ScienceFilePath.extract_filename_components(filename)
-    logger.info(f"Initial event parsed filename: {components}")
-
     with db.Session() as session:
-        # Potential jobs are the instruments that depend on the current file.
-        potential_jobs = get_downstream_dependencies(session, components)
-        logger.info(f"Potential jobs found [{len(potential_jobs)}]: {potential_jobs}")
+        # Since the SQS events can be batched together, we need to loop through
```
Comment (tech3371) [inline sds_data_manager/lambda_code/SDSCode/batch_starter.py:370]: thanks!

File: sds_data_manager/stacks/instrument_lambdas.py [hunk 12]
```diff
@@ -102,20 +105,8 @@ def __init__(
...
-            ),
-        )
-
-        event_from_indexer_lambda.add_target(
-            targets.LambdaFunction(self.instrument_lambda)
-        )
+        # This sets up the lambda to be triggered by the SQS queue. Since this is a FIFO
+        # queue, each instrument will have messages processed in order. However,
+        # different instruments will be processed in parallel, with multiple instances
+        # of the batch_starter lambda.
```
Comment (tech3371) [inline sds_data_manager/stacks/instrument_lambdas.py:111]: we can discuss about this outside of this PR. I feel like we can start with one event rule that hits SQS target and then add on as we need.

**Review by tech3371 — APPROVED** [review]
LGTM!
